### PR TITLE
Add V4 distribution implementation developer guide

### PR DESCRIPTION
**Pull request by ricardoV94**
This PR adds a developer guide on how to implement (and test) new distributions for PyMC3 `v4`. 

Somethings are still marked as `TODO` and feedback is much appreciated. 

The way distributions are implemented (and tested) will probably evolve over time, and hopefully this document will be easy to update as those things happen.

### CODE REVIEW

**Comment by ricardoV94** [comment]
@OriolAbril I need quite some help to figure out how to properly integrate this with the rest of the docs, as well things related to markdown formatting, and cross references! :) Thanks in advance

**Comment by ricardoV94** [comment]
@drabbit17, can you have a look at the section on "Adding tests for the new `RandomVariable`", to see if I missed anything obvious, or said something wrong?

File: docs/source/distribution_implementation.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+
+## 1. Creating a new `RandomVariable` `op`
+
+`RandomVariables` are responsible for implementing the random sampling methods,
+which in version 3 of PyMC3 used to be one of the standard `Distribution`
+methods, alongside `logp` and `logcdf`. The `RandomVariable` is also responsible
+for parameter broadcasting and shape inference.
+
+Before creating a new `RandomVariable` make sure that it is not offered in the
+[Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions)
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: I think this could be `` {class}`numpy.random.Generator` `` or `` {class}`Numpy library <numpy.random.generator>` `` provided we set up a numpy intershpinx mapping, which will probably come in handy in other places,

File: docs/source/distribution_implementation.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+PyMC3 distributions build on top of Aesara's random variables (TODO: add link),
+and implement `logp` and `logcdf` methods as well as other initialization and
+validation helpers, most notably `shape/dims`, alternative parametrizations, and
+default `transforms`.
+
+Here is a summary check-list of the steps needed to implement a new distribution.
+Each section will be expanded below:
+
+1. Creating a new `RandomVariable` `op`
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: general question, are there docs for RandomVariable and Distribution base classes and if so, do we want to link there?

File: docs/source/distribution_implementation.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+   distributions. Unfortunately there is no general reference documenting how
+   `size` ought to work for multivariate distributions.
+   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
+   to get more context.
+1. `Aesara` tries to infer the output shape of the `RandomVariable`
+   (given a user-specified size) by introspection of the `ndim_supp` and
+   `ndim_params` attributes. However, the default method may not work for more
+   complex distributions. In that case, custom `_shape_from_params` (and less
+   probably (?) `_infer_shape`) should also be implemented in the new
+   `RandomVariable` class. One simple example is seen in the [MultinomialRV](https://github.com/aesara-devs/aesara/blob/3a3adaee3b8ceccff6d0208996f117fcd5e5eb6f/aesara/tensor/random/basic.py#L518-L521)
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: It looks like MultinomialRV is not anywhere within Aesara's docs, once it is I think this would be better as an intersphinx link. What I don't know how it can be done is linking directly to the source code instead of link to api page that then has the `[source]` link :thinking:

File: docs/source/distribution_implementation.md [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+1. It's okay to use the `rng_fn` `classmethods` of other Aesara and PyMC3
+   `RandomVariables` inside the new `rng_fn`. For example if you are
+   implementing a negative HalfNormal `RandomVariable`, your `rng_fn` can simply
+   return `- halfnormal.rng_fn(rng, scale, size)`.
+
+
+For a quick test that your new `RandomVariable` `Op` is working, you can just
+call `eval` on a new instance:
+
+```python
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: you have structured all the code as input and output with outputs within the python highlighted code cell, would it make sense to make the file executable? and either run it every time we build docs or save it as ipynb so output gets stored and we don't need to run every time?

File: docs/source/distribution_implementation.md [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+```
+
+Some notes:
+
+1. A distribution should at the very least inherit from `Discrete` or
+   `Continuous`. For the latter, more specific subclasses exist:
+   `PositiveContinuous`, `UnitContinuous`, `BoundedContinuous`,
+   `CircularContinuous`, which specify default transformations for the variables.
+   If you need to specify a one-time custom transform you can also override the
+   `__new__` method, as is done for the [`Dirichlet`](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/multivariate.py#L399-L401).
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: This could also be `` {class}`pymc3.Dirichlet` ``, but again, this goes to the api docs that will have the `[source]` link so it's two clicks, I'll try to find a better way. 

I don't like references to github lines because they either point to a specific commit and get outdated without anyone realizing so checking the link goes to the right code but that code is no longer in use within the actual library or they point to the main/dev branch and any change in line numbers makes the link point to the right file but wrong section.

File: docs/source/distribution_implementation.md [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+1. If a distribution does not have a corresponding `random` implementation, a
+   `RandomVariable` should still be created that raises a `NotImplementedError`.
+   This is the case for the [`Flat` distribution](https://github.com/pymc-devs/pymc3/blob/6eaf9a0d7bc5b9168bedc9337808391ccb86a5ea/pymc3/distributions/continuous.py#L312-L321).
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: With FlatRV the situation is similar yet a bit different, because FlatRV _in not part of the docs!_. I understand it should not be part of the user facing docs but it looks like it's useful for dev facing docs. Would it be a good idea to have a couple of orphan API pages (therefore not indexed in the API navbar section) so we can link there from the developer guide?

File: docs/source/distribution_implementation.md [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+test_blah_logcdf(self):
+    ...
+
+
+```
+
+
+## 5. Documenting the new `Distribution`
+
+TODO: What needs to be done? Adding a new entry to the .rst?
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: At the very least, they should have a docstring written and then be referenced in the respective API page. Depending on the distribution and it's uses it should probably get a notebook example where it's used in a case study or a tutorial if it has significant difference all other distributions.

File: docs/source/distribution_implementation.md (hunk 6 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: Sounds reasonable

File: docs/source/distribution_implementation.md (hunk 3 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: The Dirichlet multinomial will have the same trick, so we can link to that #4758

File: docs/source/distribution_implementation.md (hunk 2 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: We should. We have docstrings for the distribution base classes. For the RandomVariable I am not sure if it appears in the aesara docs, but if not we can add something for sure

File: docs/source/distribution_implementation.md (hunk 4 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: I am not sure how functional the pseudo code is :/

File: docs/source/distribution_implementation.md [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+#    [0.65506507, 0.47713784],
+#    [0.61284429, 0.49720329],
+#    [0.69325978, 0.96272673]])
+
+```
+
+## 2. Inheriting from a PyMC3 base `Distribution` class
+
+After implementing the new `RandomVariable` `op`, it's time to make use of it
+in a new PyMC3 `distribution`. PyMC3 works in a very functional way, and the
```
Comment (ricardoV94) [inline docs/source/distribution_implementation.md]: We can try that glossary thing with an entry for `functional programming`

File: docs/source/distribution_implementation.md [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+
+1. Creating a new `RandomVariable` `op`
+1. Implementing the corresponding `Distribution` class
+1. Adding tests for the new `RandomVariable`
+1. Adding tests for the `logp` / `logcdf` methods
+1. Documenting the new `Distribution`.
+
+This guide does not attempt to explain the rationale behind the `Distributions`
+current implementation, and details are provided only insofar as they help
+implementing new "standard" distributions.
```
Comment (matteo-pallini) [inline docs/source/distribution_implementation.md]: >  This guide does not attempt to explain the rationale behind the `Distributions` current implementation

Would you say that [this PR](https://github.com/pymc-devs/pymc3/pull/4440) is a good reference for anyone interested in that? if so, would it make sense to add it here for anyone interested in better understanding the ratio behind the change in implementation?

**Comment by matteo-pallini** [comment]
> @drabbit17, can you have a look at the section on "Adding tests for the new RandomVariable"

Looks pretty good to me

**Comment by OriolAbril** [comment]
Also forgot to mention that this page will now be orphan, and not discoverable in the docs unless one searches for it. Not sure how but it should be added to the toctree or included within he developer guide

File: docs/source/distribution_implementation.md (hunk 9 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: After having a look, I feel it's too unstructured to be of real help. Besides it was not even merged in the end. This one contains all the changes, but it's gigantic: https://github.com/pymc-devs/pymc3/pull/4500

**Comment by ricardoV94** [comment]
> Also forgot to mention that this page will now be orphan, and not discoverable in the docs unless one searches for it. Not sure how but it should be added to the toctree or included within he developer guide

The developer guide should probably be organized hierarchically, instead of being a single file like it is now, no?

**Comment by OriolAbril** [comment]
That would probably be best, so it can also include the architecture doc, style guides...

**Comment by ricardoV94** [comment]
I updated the references to follow the format you mentioned `{class}``import`. I added some terms that would go into the glossary, let me know how I can actually add the descriptions.

**Comment by ricardoV94** [comment]
> That would probably be best, so it can also include the architecture doc, style guides...

I would defer that to a separate PR, and in the meantime just leave this guide hanging.

File: docs/source/distribution_implementation.md [hunk 10]
```diff
@@ -0,0 +1,543 @@
+# Implementing a Distribution (developer guide)
+
+This guide provides an overview on how to implement a distribution for version 4
+of PyMC3. It is designed for developers who wish to add a new distribution to
+the library. Users will not be aware of all this complexity and should instead
+make use of helper methods such as (TODO).
+
+PyMC3 {`Distributions`}`pymc3.distributions.Distribution` build on top of
+Aesara's {`RandomVariables`}`aesara.tensor.random.op.RandomVariable`, and
```
Comment (OriolAbril) [inline docs/source/distribution_implementation.md]: I now realize I did not share any reference on how intersphinx works. The `{class}` thingy indicates that we are linking to an object of type class. intersphinx docs are quite bad so I tried to document everything and add examples at https://arviz-devs.github.io/arviz/contributing/developer_guide.html#reference-external-libraries. It is still missing the custom text option though:

```
{class}`pymc3.distributions.Distribution`                 -> pymc3.distributions.Distribution
{class}`~pymc3.distribution.Distribution`                 -> Distribution
{class}`Distributions <pymc3.distribution.Distribution>`  -> Distributions
```

File: docs/source/distribution_implementation.md [hunk 11]
```diff
@@ -0,0 +1,543 @@
+# Implementing a Distribution (developer guide)
+
+This guide provides an overview on how to implement a distribution for version 4
+of PyMC3. It is designed for developers who wish to add a new distribution to
+the library. Users will not be aware of all this complexity and should instead
+make use of helper methods such as (TODO).
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]: To reduce diff complexity with markdown files, one can do one line per sentence. Single line breaks don't show up in the rendered output and with text I don't think we need to have a limit on the characters per line?

File: docs/source/distribution_implementation.md (hunk 2 above)
Comment (michaelosthege) [reply docs/source/distribution_implementation.md]: I find https://aesara.readthedocs.io/en/latest/extending/graphstructures.html to be one of the most useful chapters in the Aesara docs.

Comment (michaelosthege) [reply docs/source/distribution_implementation.md]:
```suggestion
1. Creating a new `RandomVariable` `Op`
```

File: docs/source/distribution_implementation.md [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+1. Implementing the corresponding `Distribution` class
+1. Adding tests for the new `RandomVariable`
+1. Adding tests for the `logp` / `logcdf` methods
+1. Documenting the new `Distribution`.
+
+This guide does not attempt to explain the rationale behind the `Distributions`
+current implementation, and details are provided only insofar as they help
+to implement new "standard" distributions.
+
+## 1. Creating a new `RandomVariable` `op`
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
## 1. Creating a new `RandomVariable` `Op`
```

File: docs/source/distribution_implementation.md (hunk 1 above)
Comment (michaelosthege) [reply docs/source/distribution_implementation.md]:
```suggestion
[Numpy library](https://numpy.org/doc/stable/reference/random/generator.html#distributions).
```

File: docs/source/distribution_implementation.md [hunk 13]
```diff
@@ -0,0 +1,543 @@
...
+```python
+
+from aesara.tensor.var import TensorVariable
+from aesara.tensor.random.op import RandomVariable
+
+# Create your own `RandomVariable`...
+class BlahRV(RandomVariable):
+    name: str = "blah"
+
+    # Provide the number of (output) dimensions for this RV
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
    # Provide the minimum number of (output) dimensions for this RV
```

File: docs/source/distribution_implementation.md [hunk 14]
```diff
@@ -0,0 +1,543 @@
...
+    # one a matrix and the other a scalar, `[2, 0]`; etc.)
+    ndims_params: List[int] = [0, 0]
+
+    # The NumPy/Aesara dtype for this RV (e.g. `"int32"`, `"int64"`).
+    # The standard in the library is `"int64"` for discrete variables
+    # and `"floatX"` for continuous variables
+    dtype: str = "floatX"
+
+    # A pretty text and LaTeX representation for the RV
+    _print_name: Tuple[str, str] = ("blah", "\\operatorname{blah}")
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]: Maybe choose a more meaningful name so it becomes clear how capitalization should be for these Text/LaTeX fields?

`MoonshotRV` or something.

File: docs/source/distribution_implementation.md [hunk 15]
```diff
@@ -0,0 +1,543 @@
...
+        cls,
+        rng: np.random.RandomState,
+        loc: np.ndarray,
+        scale: np.ndarray,
+        size: Tuple[int, ...],
+    ) -> np.ndarray:
+        return scipy.stats.blah.rvs(loc, scale, random_state=rng, size=size)
+
+# Create an instance of the `RandomVariable` `Op`...
+blah = BlahRV()
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]: The user API is always through things like `pm.Normal`, so what's the purpose of these instances?

File: docs/source/distribution_implementation.md [hunk 16]
```diff
@@ -0,0 +1,543 @@
...
+1. The `size` argument (together with the inputs shape) is the only way for the
+   user to specify the `batch_shape` (~number of repetitions) of the
+   `RandomVariable`.
+   The `rng_fn` will have to take this into consideration for correct output.
+   `size` is the specification used by `Numpy` and `Scipy` and works like
+   PyMC3 `shape` for univariate distributions, but is different for multivariate
+   distributions. Unfortunately there is no general reference documenting how
+   `size` ought to work for multivariate distributions.
+   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
+   to get more context.
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
1. The `size` argument (together with the inputs shape) are the one way to specify
    the dimensions (minus the last `ndim_supp`) at the `RandomVariable` API level.
    At the user API level more options (`shape`, `dims`) are available, but this is taken
    care of by {`class`}`~pymc3.distributions.Distribution`.
   The `rng_fn` will have to take `size` into consideration for correct output.
   `size` is the specification used by `Numpy` and `Scipy` and works like
   PyMC3 `shape` for univariate distributions, but is different for multivariate
   distributions. Unfortunately there is no general reference documenting how
   `size` ought to work for multivariate distributions.
   This [discussion](https://github.com/numpy/numpy/issues/17669) may be helpful
   to get more context.
```

File: docs/source/distribution_implementation.md [hunk 17]
```diff
@@ -0,0 +1,543 @@
...
+   cases, it may not be possible to make use of the default helper,
+   but those have not been found yet!
+1. It's okay to use the `rng_fn` `classmethods` of other Aesara and PyMC3
+   `RandomVariables` inside the new `rng_fn`. For example if you are
+   implementing a negative HalfNormal `RandomVariable`, your `rng_fn` can simply
+   return `- halfnormal.rng_fn(rng, scale, size)`.
+
+
+For a quick test that your new `RandomVariable` `Op` is working, you can just
+call `eval` on a new instance:
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
For a quick test that your new `RandomVariable` `Op` is working, you can call
the `Op` with parameters and call `eval()` on the resulting {class}`~aesara.tensor.var.TensorVariable` instance:
```

File: docs/source/distribution_implementation.md [hunk 18]
```diff
@@ -0,0 +1,543 @@
...
+#    [0.20748312, 0.45307477],
+#    [0.65506507, 0.47713784],
+#    [0.61284429, 0.49720329],
+#    [0.69325978, 0.96272673]])
+
+```
+
+## 2. Inheriting from a PyMC3 base `Distribution` class
+
+After implementing the new `RandomVariable` `op`, it's time to make use of it
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
After implementing the new `RandomVariable` `Op`, it's time to make use of it
```

File: docs/source/distribution_implementation.md [hunk 19]
```diff
@@ -0,0 +1,543 @@
...
+in a new PyMC3 {`Distribution`}`pymc3.distributions.Distribution`. PyMC3 works
+in a very {functional}`Functional_Programming` way, and the `distribution`
+classes are there mostly to facilitate porting the `v3` code to the new `v4`
+version, and to keep related methods organized together. In practice, they take
+care of:
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
classes are there mostly to facilitate porting the `v3` code to the new `v4`
version, add PyMC3 API features and keep related methods organized together.
In practice, they take
care of:
```

File: docs/source/distribution_implementation.md [hunk 20]
```diff
@@ -0,0 +1,543 @@
...
+   If you need to specify a one-time custom transform you can also override the
+   `__new__` method, as is done for the
+   {`Dirichlet`}`pymc3.distributions.multivariate.Dirichlet`.
+1. If a distribution does not have a corresponding `random` implementation, a
+   `RandomVariable` should still be created that raises a `NotImplementedError`.
+   This is the case for the {`Flat`}`pymc3.distributions.continuous.FlatRV`.
+   In this case it will be necessary to provide a standard `initval` by
+   overriding `__new__`.
+1. As mentioned above, `v4` works in a very {functional}`Functional_Programming`
+   way, and all the   information that is needed in the `logp` and `logcdf`
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
   way, and all the information that is needed in the `logp` and `logcdf`
```

File: docs/source/distribution_implementation.md [hunk 21]
```diff
@@ -0,0 +1,543 @@
...
+   methods is expected to be "carried" via the `RandomVariable` inputs. You may
+   pass numerical arguments that are not strictly needed for the `random` method
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
   pass numerical arguments that are not strictly needed for the `rng_fn` method
```
?

File: docs/source/distribution_implementation.md [hunk 22]
```diff
@@ -0,0 +1,543 @@
...
+```python
+sizes_to_check = [None, (1), (2, 3)]
+sizes_expected = [(3,), (1, 3), (2, 3, 3)]
+tests_to_run = ["check_rv_size"]
+```
+
+This is usually needed for Multivariate distributions. You can see an example in
+the {`Dirichlet` test}`pymc3.test.test_random.TestDirichlet`.
+
+### Notes on `check_pymcs_draws_match_reference` test
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]: Maybe move this into a docstring right at the test definition?

File: docs/source/distribution_implementation.md [hunk 23]
```diff
@@ -0,0 +1,543 @@
...
+   not very informative about the correctness of the implementation).
+1. By default, a random subset of 100 `param` * `paramdomain` combinations is
+   tested, in order to keep the test runtime under control. When testing your
+   shiny new distribution, you can temporarily set `n_samples=-1` to force all
+   combinations to be tested. This is important to avoid the your `PR` leading
+   to surprising failures in future runs whenever some bad combinations of
+   parameters are randomly tested.
+1. On Github all tests are run twice, under the `aesara.config.floatX` flags of
+   `"float64"` and `"float32"`. However, the reference Python functions will
+   run in a pure "float64" environment, which means the reference and the `Pymc3`
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
   run in a pure "float64" environment, which means the reference and the `PyMC3`
```

File: docs/source/distribution_implementation.md [hunk 24]
```diff
@@ -0,0 +1,543 @@
...
+
+    f(x \mid \param1, \param2) = \exp{x * (param1 + \log{param2})}
+
+ .. plot::
+
+     import matplotlib.pyplot as plt
+     import numpy as np
+     import scipy.stats as st
+     import arviz as az
+     plt.style.use('arviz-darkgrid')
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]:
```suggestion
```
These style changes are unrelated to the code that's being demonstrated. If we consider this important for the documentation we should set the style at docs-build-time.

**Review by michaelosthege — CHANGES_REQUESTED** [review]
Looks almost complete to me!
Let's address the ToDos and merge it sooner than later.

File: docs/source/distribution_implementation.md (hunk 15 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: This is what actually creates the `RandomVariable Op`.

File: docs/source/distribution_implementation.md (hunk 14 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: What would that show differently, underscores?

File: docs/source/distribution_implementation.md (hunk 11 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: Done

File: docs/source/distribution_implementation.md (hunk 15 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: Added a more clear comment, not sure if I am saying something incorrect

File: docs/source/distribution_implementation.md (hunk 16 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: Added a slightly different note on this below.

File: docs/source/distribution_implementation.md (hunk 14 above)
Comment (michaelosthege) [reply docs/source/distribution_implementation.md]: I'm asking because the LaTeX operators for `Uniform`, or `Normal` start with capital letters, right? But this example does not...

Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: Ahhh... I'll change it then. Missed that one

File: docs/source/distribution_implementation.md [hunk 25]
```diff
@@ -0,0 +1,425 @@
+# Implementing a Distribution (developer guide)
+
+This guide provides an overview on how to implement a distribution for version 4 of PyMC3.
+It is designed for developers who wish to add a new distribution to the library.
+Users will not be aware of all this complexity and should instead make use of helper methods such as (TODO).
```
Comment (michaelosthege) [inline docs/source/distribution_implementation.md]: Leftover TODO

**Review by michaelosthege — CHANGES_REQUESTED** [review]

File: docs/source/distribution_implementation.md (hunk 25 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]: It's still a pymc3 ToDo. This was the role of `Densitydist` but that was deprecated or at least not refactored for V4. Maybe we can link to something like: https://github.com/pymc-devs/pymc-examples/pull/185

**Comment by zoj613** [comment]
I think adding information about extending `Op` for new distributions would be valuable. Aesara contains little to no information about this. The example at https://aesara.readthedocs.io/en/latest/extending/extending_aesara.html#example-op-definition is unhelpful and seems very incomplete, nor does it ever touch on the "optional" methods. Moreover, details about dealing with shapes is completely missing. It has been quite frustrating for me to plough through multiple pages of the docs and still not get it.

**Comment by ricardoV94** [comment]
> I think adding information about extending `Op` for new distributions would be valuable. Aesara contains little to no information about this. The example at https://aesara.readthedocs.io/en/latest/extending/extending_aesara.html#example-op-definition is unhelpful and seems very incomplete, nor does it ever touch on the "optional" methods. Moreover, details about dealing with shapes is completely missing. It has been quite frustrating for me to plough through multiple pages of the docs and still not get it.

Do you want to open a Github issue for this? I think it could be important, but perhaps out of scope for this guide.

File: docs/source/distribution_implementation.md (hunk 14 above)
Comment (ricardoV94) [reply docs/source/distribution_implementation.md]:
```suggestion
    _print_name: Tuple[str, str] = ("Blah", "\\operatorname{Blah}")
```

**Comment by ricardoV94** [comment]
The TODO may point to https://github.com/pymc-devs/pymc-examples/pull/185 in the future, but that is still an early WIP. I think it's fine to leave it for the time-being. Similarly, the glossary will not work until #4797 is wrapped up. 

I am okay merging this and leaving it as an orphan-page until the "dev-branch" that @martinacantaro has suggested is created. If not, we can leave this as a draft PR. As you guys see fit.

**Review by michaelosthege — APPROVED** [review]
